File: csharpcore/GildedRose.cs
using System.Collections.Generic;
using System.Linq;

namespace csharpcore
{
    public class GildedRose
    {
        private readonly Dictionary<string, IItemCategory> _categories;
        private readonly IItemCategory _defaultCategory;
        private readonly List<Item> _items;

        public GildedRose(IEnumerable<Item> items, IItemCategory defaultCategory, params IItemCategory[] categories)
        {
            _defaultCategory = defaultCategory;
            _items = items.ToList();
            _categories = categories.ToDictionary(c => c.Name);
        }

        public void UpdateQuality()
        {
            foreach (var item in _items)
            {
                if (_categories.TryGetValue(item.Name, out var category))
                {
                    category.UpdateSellIn(item);
                    category.UpdateQuality(item);
                }
                else
                {
                    _defaultCategory.UpdateSellIn(item);
                    _defaultCategory.UpdateQuality(item);
                }
            }
        }
    }
}

File: Java/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

import java.util.HashMap;
import java.util.Map;

public class GildedRose {

    private Map<String, ItemCategory> categories;
    private ItemCategory defaultCategory;
    private Item[] items;

    public GildedRose(
        Item[] items,
        ItemCategory defaultCategory,
        ItemCategory... categories
    ) {
        this.items = items;
        this.defaultCategory = defaultCategory;

        this.categories = new HashMap<String, ItemCategory>();
        for (int i = 0; i < categories.length; i++) {
            this.categories.put(categories[i].getName(), categories[i]);
        }
    }

    public Item[] getItems() {
        return this.items;
    }

    public void updateQuality() {
        for (Item item : items) {
            ItemCategory category = this.categories.containsKey(item.name)
                ? this.categories.get(item.name)
                : this.defaultCategory;

            category.UpdateSellIn(item);
            category.UpdateQuality(item);
        }
    }
}
